### PR TITLE
[onnx.export] Cache SetGraphInputTypeReliable

### PR DESCRIPTION
**Pull request by gustavla**
This PR is part of an effort to speed up torch.onnx.export (https://github.com/pytorch/pytorch/issues/121422).

- For each node that is processed in onnx.export, a check is run to see if all inputs are "reliable" (static shape, etc.). This value does not change, so it is much faster to cache it on the first computation. The caching is added to the ConstantMap state.
- Resolves (6) in #121422.
- Also see #123028 with a similar addition of a cache state.

(partial fix of #121545)

### CODE REVIEW

**Review by justinchuby — APPROVED** [review]
Like other PRs I would include the motivation of the change as comments

**Comment by gustavla** [comment]
> Like other PRs I would include the motivation of the change as comments

@justinchuby Motivation added as part of comment. Thanks!

**Comment by gustavla** [comment]
> Failing merge rule: Core Maintainers

@justinchuby Looks like I can't run the merge command, so I'll need to bug you about it. 🙇

**Comment by gustavla** [comment]
@justinchuby This is the final one now and it just needs a nudge. Thank you so much for all the help reviewing and merging these.

File: third_party/kineto
Comment (justinchuby) [inline third_party/kineto:1]: Could you revert this change? Thanks

Comment (gustavla) [reply third_party/kineto:1]: Nice catch! Sorry that slipped in. I've rebased and pushed.

Comment (gustavla) [reply third_party/kineto:1]: I rebased onto main. Not sure if I should have kept it against `refs/remotes/origin/viable/strict`.

**Comment by gustavla** [comment]
I'm not sure why the merge failed (intermittent failure?). Everything looks green and good to merge, but the merge got cancelled.

**Comment by gustavla** [comment]
@justinchuby Good to merge?
